File: heap/leetcode_2099/solution.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSubsequence = function(nums, k) {
    const indexedNums = nums.map((num, index) => [num, index]);

    indexedNums.sort((a, b) => b[0] - a[0]);

    const topK = indexedNums.slice(0, k).sort((a, b) => a[1] - b[1]);
    const result = topK.map(pair => pair[0]);

    return result;
};

File: heap/leetcode_2099/solution.py
from typing import List
import heapq

class Solution:
    def maxSubsequence(self, nums: List[int], k: int) -> List[int]:
        top_k = heapq.nlargest(k, enumerate(nums), key=lambda x: x[1])

        top_k.sort(key=lambda x: x[0])

        return [val for idx, val in top_k]